### ISSUE TITLE
Improvement request - please import Waypoint/POI 

### ISSUE THREAD

**Issue by bwims** [pdhoopr/gps-to-gpx#15]
Hi, this is a wonderful tool for importing tracks, but it would be even better if it could also highlight individual POIs with identifying icons/text.  Would that be hard to implement?

I was wondering if the input array could have another element indicating that it was a POI rather than a point on a track, e.g. "name" and it could generate a "wpt" instead of a "trk".

Thanks!

**Comment by pdhoopr** [comment]
Hi there,

I think I get the gist of what you're asking, but I'd just like to make sure I'm totally clear!

Right now, the tool just assumes all waypoints are part of a track, so you get something like this (as you've pointed it):

```
<trk>
  <trkseg>
    <trkpt></trkpt>
    <trkpt></trkpt>
    <trkpt></trkpt>
  </trkseg>
</trk>
```

Are you looking for the ability to just ungroup these into a flat list of waypoints? Like so:

```
<wpt></wpt>
<wpt></wpt>
<wpt></wpt>
```

Or are you hoping to mix and match the two for individual points? Like so:

```
<wpt></wpt>
<wpt></wpt>
<wpt></wpt>
<trk>
  <trkseg>
    <trkpt></trkpt>
    <trkpt></trkpt>
    <trkpt></trkpt>
  </trkseg>
</trk>
```

Just curious because I've been wanting to rewrite a v2 of the library for years now and am just now starting to plan out what needs to change. I'd like to consider what you're saying and see what I can do to implement it in the next major version.

Thank you!

**Comment by bwims** [comment]
Hi!

It's the "mix and match" option really.  I use Owntracks and it provides the ability to mark the current location as a waypoint/POI.  It uses MQTT to record points as location (track) or waypoint (POI).   Ideally, since you use an array called waypoints to generate the trackpoints, you'd need a way to discriminate an array of waypoints - maybe an array called POI?

Also, instead of fields
      "latitude": 26.852324,
      "longitude": -80.08045,
      "elevation": 0,
      "time": "2016-07-06T12:36:00Z"

You'd need 

Also, instead of fields
      "latitude": 26.852324,
      "longitude": -80.08045,
      "name": "My POI",
      "time": "2016-07-06T12:36:00Z"

In practice, on a personal basis, I'd envisage either converting tracks OR POIs into GPX format.

Incidentally, if you are into more improvements, there is also a speed field that I could populate from the velocity field produced by Owntracks.    The full range of data supplied is [here](https://owntracks.org/booklet/tech/json/#_typelocation)

Thanks for being so helpful!

**Comment by pdhoopr** [comment]
Okay, that makes sense. Thanks for the additional information! I imagine this is completely reasonable and will give it some thought.

Regarding the speed field: that's a bit of a gray area in the current version of the library. In fact, you _can_ provide a `speedKey` and it will output a `<speed>` element, e.g.:

```js
createGpx(waypoints, {
  speedKey: 'vel'
});
```

This is undocumented, though, because I realized later that `<speed>` is only supported in GPX v1.0 and this library outputs v1.1. So the GPX that gets created is probably technically invalid. Knowing this, you're obviously welcome to use the option if you want to and don't mind GPX that isn't fully according to spec.

**Comment by pdhoopr** [comment]
Alternatively, I think the proper way to add speed in GPX 1.1 is via extensions on each point. Unfortunately, that would probably mean manipulating your data before you call the function:

```js
const waypointsWithExtensions = waypoints.map(wpt => ({
  ...wpt,
  extensions: {
    speed: wpt.vel
  }
}))
createGpx(waypointsWithExtensions)
```

This is also supported in the current version of the library. Although I think the XML namespacing is a little off so may not be fully valid, but certainly closer than usually an unsupported tag from the older 1.0 spec.

Sorry for the long explanations! However, you should be able to get speed as part of your output one way or another.

**Comment by bwims** [comment]
Well, I have to manipulate the data anyway to get it into the form required; it looks like this:

{"_type":"location","acc":4,"alt":108,"batt":76,"conn":"m","inregions":["Home"],"lat":51.2004176,"lon":1.2638549,"tid":"BW","tst":1585417653,"vac":2,"vel":4}
{"_type":"location","acc":6,"alt":106,"batt":75,"conn":"m","inregions":["Home"],"lat":51.2007667,"lon":1.264426,"tid":"BW","tst":1585417684,"vac":2,"vel":3}
{"_type":"location","acc":4,"alt":106,"batt":75,"conn":"m","inregions":["Home"],"lat":51.2009216,"lon":1.2649884,"tid":"BW","tst":1585417713,"vac":2,"vel":3}
{"_type":"location","acc":8,"alt":103,"batt":75,"conn":"m","inregions":["Home"],"lat":51.2011551,"lon":1.2653781,"tid":"BW","tst":1585417743,"vac":2,"vel":3}
{"_type":"location","acc":18,"alt":103,"batt":75,"conn":"m","inregions":["Home"],"lat":51.2014123,"lon":1.26616,"tid":"BW","tst":1585417774,"vac":2,"vel":4}

So manipulating that to create some extra JSON fields wouldn't be a problem.

I presume the extensions field is something you are proposing for V2 ? 
If so, I'd rather go for something that's 1.1 compatible.

Thanks again!

**Comment by pdhoopr** [comment]
Actually extension support exists in the current version, so you could use some form of the code snippet I sent above to output your velocity to a `<speed>` element. I think the namespace declarations on the GPX element itself are a little off, though, so it's not a perfect solution.

v1 hardcodes Garmin's extensions for trackpoints, but v2 will probably be bring-your-own-extensions, so the way to add them will change, too. Completely up to you what you'd like to do for now!

Wanted to clarify one last point for your POI use case. Are you interested in having POIs be part of the same GPX output as tracks [example 1 below], or are separate GPX elements okay [example 2 below]?

**Example 1: single GPX element**

```xml
<gpx>
  <wpt></wpt>
  <wpt></wpt>
  <wpt></wpt>

  <trk>
    <trkseg>
      <trkpt></trkpt>
      <trkpt></trkpt>
    </trkseg>
  </trk>
</gpx>
```

**Example 2: multiple GPX elements**

```xml
<gpx>
  <wpt></wpt>
  <wpt></wpt>
  <wpt></wpt>
</gpx>

<gpx>
  <trk>
    <trkseg>
      <trkpt></trkpt>
      <trkpt></trkpt>
    </trkseg>
  </trk>
</gpx>
```

**Comment by pdhoopr** [comment]
On second thought, I think I'm going to fix the extensions namespacing issue. I'll follow back up here once I release a fix, at which point you should feel free to use extensions in v1 without hesitation.

**Comment by bwims** [comment]
Whichever is easiest for you, Patrick.  I can separate the waypoints out
easily enough.

**Comment by bwims** [comment]
Fine with me.  All I really need is some more documentation on how to use
the features already supported.  I don't think your example on the
README.MD includes extensions.  Is there another document somewhere?

Thanks!

**Comment by pdhoopr** [comment]
Yep, you're definitely right; extension usage is not documented anywhere. I'm going to try and get some fixes out today and will include that documentation as part of that. I'll let you know once it's published.

Thanks!

**Comment by bwims** [comment]
You are a star!

**Comment by pdhoopr** [comment]
Alright, please check out [this section about extensions](https://github.com/pdhoopr/gps-to-gpx#extensions) in the README and give v1.5.1 a try. Let me know how it goes!
